src/shared/yagna/adapters/market-api-adapter.ts

```diff
@@ -150,4 +213,184 @@ export class MarketApiAdapter implements MarketApi {
+          new Observable<AgreementEvent>((observer) => {
+            try {
+              this.logger.debug("Market API Adapter received agreement event", { event });
+
+              const timestamp = new Date(Date.parse(event.eventDate));
+
+              // @ts-expect-error FIXME #yagna, wasn't this fixed? {@issue https://github.com/golemfactory/yagna/pull/3136}
+              const eventType = event.eventType || event.eventtype;
+
+              this.getAgreement(event.agreementId)
+                .then((agreement) => {
+                  switch (eventType) {
+                    case "AgreementApprovedEvent":
+                      observer.next({
+                        type: "AgreementApproved",
+                        agreement,
+                        timestamp,
+                      });
+                      break;
+                    case "AgreementTerminatedEvent":
+                      observer.next({
+                        type: "AgreementTerminated",
+                        agreement,
+                        // @ts-expect-error FIXME #ya-ts-client event type discrimination doesn't work nicely with the current generator
+                        terminatedBy: event.terminator,
+                        // @ts-expect-error FIXME #ya-ts-client event type discrimination doesn't work nicely with the current generator
+                        reason: event.reason.message,
+                        timestamp,
+                      });
+                      break;
+                    case "AgreementRejectedEvent":
+                      observer.next({
+                        type: "AgreementRejected",
+                        agreement,
+                        // @ts-expect-error FIXME #ya-ts-client event type discrimination doesn't work nicely with the current generator
+                        reason: event.reason.message,
+                        timestamp,
+                      });
+                      break;
+                    case "AgreementCancelledEvent":
+                      observer.next({
+                        type: "AgreementCancelled",
+                        agreement,
+                        timestamp,
+                      });
+                      break;
+                    default:
+                      this.logger.warn("Unsupported agreement event type for event", { event });
+                      break;
+                  }
+                })
+                .catch((err) => this.logger.error("Failed to load agreement", { agreementId: event.agreementId, err }));
+            } catch (err) {
+              const golemMarketError = new GolemMarketError(
+                "Error while processing agreement event from yagna",
+                MarketErrorCode.InternalError,
+                err,
+              );
+              this.logger.error(golemMarketError.message, { event, err });
+              observer.error(err);
+            }
+          }),
+      ),
+    );
+  }
```

Comment:
`.pipe(map())`

Reply:
we use Promises inside and I think the switchMap will be better after all..